### PR TITLE
Preloads more nvidia pypi library for multi arch distributions

### PR DESCRIPTION
**Pull request by syed-ahmed**
Following the same logic of preloading cudnn and cublas from the pypi folder in multi-arch disributions, where Pure-lib vs Plat-lib matters, this PR adds the logic for the rest of the cuda pypi libraries that were integrated.

I have tested this PR by running the code block locally and installing/uninstalling nvidia pypi libraries:

```
import sys
import os

def _preload_cuda_deps():
    """Preloads cudnn/cublas deps if they could not be found otherwise."""
    # Should only be called on Linux if default path resolution have failed

    cuda_libs = {
        'cublas': 'libcublas.so.11',
        'cudnn': 'libcudnn.so.8',
        'cuda_nvrtc': 'libnvrtc.so.11.2',
        'cuda_runtime': 'libcudart.so.11.0',
        'cuda_cupti': 'libcupti.so.11.7',
        'cufft': 'libcufft.so.10',
        'curand': 'libcurand.so.10',
        'cusolver': 'libcusolver.so.11',
        'cusparse': 'libcusparse.so.11',
        'nccl': 'libnccl.so.2',
        'nvtx': 'libnvToolsExt.so.1',
    }
    cuda_libs_paths = {lib_folder: None for lib_folder in cuda_libs.keys()}

    for path in sys.path:
        nvidia_path = os.path.join(path, 'nvidia')
        if not os.path.exists(nvidia_path):
            continue
        for lib_folder, lib_name in cuda_libs.items():
            candidate_path = os.path.join(nvidia_path, lib_folder, 'lib', lib_name)
            if os.path.exists(candidate_path) and not cuda_libs_paths[lib_folder]:
                cuda_libs_paths[lib_folder] = candidate_path
        if all(cuda_libs_paths.values()):
            break
    if not all(cuda_libs_paths.values()):
        none_libs = [lib for lib in cuda_libs_paths if not cuda_libs_paths[lib]]
        raise ValueError(f"{', '.join(none_libs)} not found in the system path {sys.path}")


_preload_cuda_deps()
```

I don't have access to a multi-arch environment, so if somebody could verify a wheel with this patch on a multi-arch distribution, that would be great!

cc: @atalman @malfet @ptrblck

### CODE REVIEW

**Comment by ezyang** [comment]
Tagging in experts, but hard coding version numbers looks questionable

**Comment by atalman** [comment]
> Tagging in experts, but hard coding version numbers looks questionable

Agree with @ezyang . @syed-ahmed maybe there is a way to hardcode only major versions ? 
Something along the lines: ``libnvrtc.so.11.*``

**Comment by atalman** [comment]
@syed-ahmed @malfet 

I run some tests today with Amazon linux 2. I can't reproduce this issue with python3.8(installed) and up python3.9 (compiled):
3.8 + torch installed installed via:
```
sudo amazon-linux-extras enable python3.8
yum install -y python38 python38-devel
python3.8 -m pip install torch-2.0.0.dev20230210%2Bcu117.with.pypi.cudnn-cp38-cp38-linux_x86_64.whl 
```

I see all the libraries are installed in same folder:
```
ls /home/ec2-user/.local/lib/python3.8/site-packages/
filelock                                 nvidia_cuda_cupti_cu11-11.7.101.dist-info   sympy
filelock-3.9.0.dist-info                 nvidia_cuda_nvrtc_cu11-11.7.99.dist-info    sympy-1.11.1.dist-info
functorch                                nvidia_cuda_runtime_cu11-11.7.99.dist-info  torch
isympy.py                                nvidia_cudnn_cu11-8.5.0.96.dist-info        torch-2.0.0.dev20230210+cu117.with.pypi.cudnn.dist-info
mpmath                                   nvidia_cufft_cu11-10.9.0.58.dist-info       torchgen
mpmath-1.2.1.dist-info                   nvidia_curand_cu11-10.2.10.91.dist-info     typing_extensions-4.4.0.dist-info
networkx                                 nvidia_cusolver_cu11-11.4.0.1.dist-info     typing_extensions.py
networkx-3.0.dist-info                   nvidia_cusparse_cu11-11.7.4.91.dist-info    wheel
nvfuser                                  nvidia_nccl_cu11-2.14.3.dist-info           wheel-0.38.4.dist-info
nvidia                                   nvidia_nvtx_cu11-11.7.91.dist-info
nvidia_cublas_cu11-11.10.3.66.dist-info  __pycache__
```

**Review by atalman — CHANGES_REQUESTED** [review]
On line 192 
```
if 'libcublas.so.11' not in err.args[0]:
```
Its looking for particular cuda libarary and this is fragile (since after recent modifications error is regarding libcurand.so.10), we would need to look for any cuda library in our list of cuda libraries:

We need changes similar to this:
```
 for lib_folder, lib_name in cuda_libs.items():
            if lib_name in err.args[0]:
                _preload_cuda_deps()
                ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
                return


 raise err
```
And declare the cuda_libs outside of the _preload_cuda_deps() method

File: torch/__init__.py [hunk 1]
```diff
@@ -141,25 +141,38 @@ def _preload_cuda_deps():
     """Preloads cudnn/cublas deps if they could not be found otherwise."""
     # Should only be called on Linux if default path resolution have failed
     assert platform.system() == 'Linux', 'Should only be called on Linux'
-    cublas_path = None
-    cudnn_path = None
+
+    cuda_libs: Dict[str, str] = {
+        'cublas': 'libcublas.so.11',
+        'cudnn': 'libcudnn.so.8',
+        'cuda_nvrtc': 'libnvrtc.so.11.2',
```
Comment (malfet) [inline torch/__init__.py:148]: Sorry, why do we need to preload nvrtc here? (Aren't it supposed to be loaded lazily anyway)

File: torch/__init__.py [hunk 2]
```diff
@@ -141,25 +141,38 @@ def _preload_cuda_deps():
...
+        'cuda_runtime': 'libcudart.so.11.0',
+        'cuda_cupti': 'libcupti.so.11.7',
+        'cufft': 'libcufft.so.10',
+        'curand': 'libcurand.so.10',
+        'cusolver': 'libcusolver.so.11',
+        'cusparse': 'libcusparse.so.11',
+        'nccl': 'libnccl.so.2',
+        'nvtx': 'libnvToolsExt.so.1',
```
Comment (malfet) [inline torch/__init__.py:156]: Correct me if I'm wrong, but aren't we supposed to move to header-only nvtx in this realease?
```suggestion
```

File: torch/__init__.py [hunk 3]
```diff
@@ -141,25 +141,38 @@ def _preload_cuda_deps():
...
+        'cublas': 'libcublas.so.11',
+        'cudnn': 'libcudnn.so.8',
+        'cuda_nvrtc': 'libnvrtc.so.11.2',
+        'cuda_runtime': 'libcudart.so.11.0',
+        'cuda_cupti': 'libcupti.so.11.7',
+        'cufft': 'libcufft.so.10',
+        'curand': 'libcurand.so.10',
+        'cusolver': 'libcusolver.so.11',
+        'cusparse': 'libcusparse.so.11',
+        'nccl': 'libnccl.so.2',
```
Comment (malfet) [inline torch/__init__.py:155]: Perhaps I've missed a prep change, but I thought PyTorch still builds nccl from source, isn't it?

**Review by malfet — CHANGES_REQUESTED** [review]
Please include testplan as well as link to corresponding change that adds all those libraries as dynamic dependencies.
And +1 on @ezyang idea of using globs rather than hardcoded versions

File: torch/__init__.py (hunk 3 above)
Comment (atalman) [reply torch/__init__.py:155]: @malfet This is nccl change in builder that sets a dynamic version: https://github.com/pytorch/builder/pull/1184

File: torch/__init__.py (hunk 1 above)
Comment (atalman) [reply torch/__init__.py:148]: @syed-ahmed @malfet I think lazy loading was enabled by this PR if I am not mistaken: https://github.com/pytorch/pytorch/pull/85692 . However the question, will lazy loading will find this dll if we don't preload ?

Comment (syed-ahmed) [reply torch/__init__.py:148]: libnvrtc.so.11.2 is listed as a dependency for libcaffe2_nvrtc:
```
ldd libcaffe2_nvrtc.so
        linux-vdso.so.1 (0x00007ffe179d6000)
        libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007fcb8a61d000)
        libnvrtc.so.11.2 => not found
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fcb8a294000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcb89ef6000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fcb89cde000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcb898ed000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fcb896e9000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fcb894ca000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fcb892c2000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fcb8ba81000)
```
Also @atalman's concern is valid I think. Where is nvrtc lazily loaded in pytorch? (I don't think [CUDA_MODULE_LOADING](https://docs.nvidia.com/cuda/cuda-c-programming-guide/lazy-loading.html?highlight=cuda_module_loading#what-is-lazy-loading) has anything to do with lazy loading of dlls, and it's for lazy loading of kernels)

File: torch/__init__.py (hunk 2 above)
Comment (syed-ahmed) [reply torch/__init__.py:156]: I see it listed as a dependency for libtorch.so. Was there a commit for header-only nvtx?

**Comment by syed-ahmed** [comment]
I've changed the hardcoded versions to be wildcards in glob. Additionally fixed the error handling @atalman pointed out.

Links to PRs that added the deps: https://github.com/pytorch/pytorch/pull/89944, https://github.com/pytorch/builder/pull/1248

Testplan:
I tested the diff as follows:
- since I don't have access to multi arch, I just moved the nvidia folder from site-package to `/miniconda3/lib/python3.8/` (which is also in sys.path) to trigger the code path in this diff. `import torch` successfully worked.
- `import torch` also works for a vanilla install of the wheel (nvidia packages in `site-packages`).
- I ran the following toy ddp example successfully with the nvidia packages not in `site-packages` folder, on a multi-gpu machine: `torchrun --nproc_per_node=2 launch.py`
```
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim

from torch.nn.parallel import DistributedDataParallel as DDP

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))


def demo_basic():
    dist.init_process_group("nccl")
    rank = dist.get_rank()
    print(f"Start running basic DDP example on rank {rank}.")

    # create model and move it to GPU with id rank
    device_id = rank % torch.cuda.device_count()
    model = ToyModel().to(device_id)
    ddp_model = DDP(model, device_ids=[device_id])

    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)

    optimizer.zero_grad()
    outputs = ddp_model(torch.randn(20, 10))
    labels = torch.randn(20, 5).to(device_id)
    loss_fn(outputs, labels).backward()
    optimizer.step()

if __name__ == "__main__":
    demo_basic()
```

**Review by atalman — APPROVED** [review]
Validated on Amazon Linux 2. Works well! Thank you for refactoring Syed!

File: torch/__init__.py (hunk 2 above)
Comment (malfet) [reply torch/__init__.py]: https://github.com/pytorch/pytorch/pull/90689

**Comment by syed-ahmed** [comment]
I started a pipeline in our nightly CI with the wheel generated from this PR and vision, timm and nvidia dl examples benchmarks passed.

**Pull request closed without merging**